core/.js/src/main/scala/io/udash/routing/UrlChangeProvider.scala

```diff
@@ -91,28 +93,21 @@ object WindowUrlPathChangeProvider extends UrlChangeProvider {
 
   override def initialize(): Unit = {
     window.document.addEventListener("click", (event: MouseEvent) => {
-      def findLink(el: Node): Element =
-        if (el == null) null
-        else if (el.nodeName.toLowerCase() == "a") el.asInstanceOf[Element]
-        else findLink(el.parentNode)
-
-      val target = findLink(event.target.asInstanceOf[Node])
-      if (target != null) {
-        val href = target.getAttribute("href")
-        val location = window.location
-        val newUrl = new JSUrl(href, location.toString)
-        val (samePath, sameHash, sameOrigin) =
-          (isSamePath(location, newUrl), isSameHash(location, newUrl), isSameOrigin(location, newUrl))
-        val ignore = shouldIgnoreClick(event, target, href, samePath, sameHash, sameOrigin)
-
-        if (!ignore) {
-          if (!samePath) {
-            val url = Url(href)
-            changeFragment(url)
+      event.target.opt
+        .collect { case node: Node => node }
+        .flatMap(Iterator.iterate(_)(_.parentNode).takeWhile(_ != null).collectFirstOpt { case a: HTMLAnchorElement => a })
+        .filter(_.href != null)
+        .foreach { target =>
+          val href = target.getAttribute("href")
+          val location = window.location
+          val newUrl = new JSUrl(href, location.toString)
+          val (samePath, sameHash, sameOrigin) =
+            (isSamePath(location, newUrl), isSameHash(location, newUrl), isSameOrigin(location, newUrl))
+          if (!shouldIgnoreClick(event, target, href, samePath, sameHash, sameOrigin)) {
+            if (!samePath) changeFragment(Url(href))
+            event.preventDefault()
```

Comment:
should we also `stopPropagation()` here?

Reply:
Do you see a reason to? It can only prevent user callbacks from execution on this afaik and as a framework we shouldn't prevent elaborate hacks unless we really need to (as is the case with preventing browser default action).

Reply:
After some thought I think it's fine as it is now.